### PR TITLE
[SPARK-35661][SQL] Allow deserialized off-heap memory entry

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/32534

#32534 proposed a use case to use `DeserializedMemoryEntry` to store off-heap data, and let Spark release the memory via the `AutoCloseable` interface. However, there is one more problem: `DeserializedMemoryEntry` always reports its size as on-heap size, which is inaccurate. If the Spark cluster is configured with small on-heap size and large off-heap size, this will trigger a lot of spilling.

This PR makes `DeserializedMemoryEntry` truly support off-heap data. Now the caller side can cache off-heap data with a new storage level `OFF_HEAP_ONLY_DESER`.

### Why are the changes needed?

correct the memory counting for off-heap data.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

updated test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @xuechendi @srowen

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 1]
```diff
@@ -474,15 +475,15 @@ private[spark] class MemoryStore(
             // non-blocking "tryLock" here in order to ignore blocks which are locked for reading:
             if (blockInfoManager.lockForWriting(blockId, blocking = false).isDefined) {
               selectedBlocks += blockId
-              freedMemory += pair.getValue.size
+              freedMemory += entry.size
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:478]: A small cleanup. We have `val entry = pair.getValue` a few lines above.

**Comment by cloud-fan** [comment]
also cc @viirya @Ngone51

File: core/src/main/java/org/apache/spark/api/java/StorageLevels.java [hunk 2]
```diff
@@ -36,6 +36,7 @@
   public static final StorageLevel MEMORY_AND_DISK_SER = create(true, true, false, false, 1);
   public static final StorageLevel MEMORY_AND_DISK_SER_2 = create(true, true, false, false, 2);
   public static final StorageLevel OFF_HEAP = create(true, true, true, false, 1);
+  public static final StorageLevel OFF_HEAP_ONLY_DESER = create(false, true, true, true, 1);
```
Comment (srowen) [inline core/src/main/java/org/apache/spark/api/java/StorageLevels.java:39]: Do we need to document this anywhere else or is this kind or private?
Seems reasonable to me.

Comment (cloud-fan) [reply core/src/main/java/org/apache/spark/api/java/StorageLevels.java:39]: good point. Maybe I should remove it and the vendors should construct this storage level by themselves, if they have deserialized off-heap entries.

It's not a user-facing thing.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 3]
```diff
@@ -706,6 +709,7 @@ private class SerializedValuesHolder[T](
     classTag: ClassTag[T],
     memoryMode: MemoryMode,
     serializerManager: SerializerManager) extends ValuesHolder[T] {
+
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:712]: nit: unnecessary blank?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 4]
```diff
@@ -298,10 +298,12 @@ private[spark] class BlockManager(
 
     protected def saveToDiskStore(): Unit
 
-    private def saveDeserializedValuesToMemoryStore(inputStream: InputStream): Boolean = {
+    private def saveDeserializedValuesToMemoryStore(
+        inputStream: InputStream,
+        memoryMode: MemoryMode): Boolean = {
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:303]: nit: seems no need to pass `memoryMode`, just follow `level.memoryMode`.

**Review by viirya — COMMENTED** [review]
Looks good to me. Do we have some context about why we don't support off-heap for deserialized storage before?

**Comment by cloud-fan** [comment]
> Do we have some context about why we don't support off-heap for deserialized storage before?

I'm not very sure but it seems like just a missed use case.

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**